src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h

```diff
@@ -1682,7 +1772,176 @@ namespace UiaOperationAbstraction
         }
     }
 
-    class UiaCacheRequest: public UiaTypeBase<
+    template <class... ItemWrapperType>
+    class UiaTuple : public UiaTypeBase<
+        std::shared_ptr<std::tuple<typename ItemWrapperType::LocalType...>>,
+        winrt::Microsoft::UI::UIAutomation::AutomationRemoteArray>
+    {
+    public:
+        using TupleLocalType = std::tuple<typename ItemWrapperType::LocalType...>;
+
+        template<std::size_t I>
+        using TupleItemWrapperType = typename std::tuple_element<I, std::tuple<ItemWrapperType...>>::type;
+
+        static constexpr auto c_anyTest = &winrt::Microsoft::UI::UIAutomation::AutomationRemoteAnyObject::IsArray;
+        static constexpr auto c_anyCast = &winrt::Microsoft::UI::UIAutomation::AutomationRemoteAnyObject::AsArray;
+
+        UiaTuple() : UiaTuple(TupleLocalType{})
+        {
+        }
+
+        UiaTuple(TupleLocalType tuple) :
+            UiaTypeBase(std::make_shared<TupleLocalType>(std::move(tuple)))
+        {
+            ToRemote();
+        }
+
+        UiaTuple(const std::shared_ptr<TupleLocalType>& value) : UiaTypeBase(details::GetOrMakeSharedPtr(value))
+        {
+            ToRemote();
+        }
+
+        UiaTuple(typename ItemWrapperType::LocalType... args) : UiaTuple(TupleLocalType{ std::move(args)... })
+        {
+        }
+
+        UiaTuple(ItemWrapperType... args) : UiaTuple()

```

Comment:
So I left it like this yesterday, but this is a bit inefficient, because:
1. It creates a local `tuple`,
2. It then `ToRemote`s it (which basically creates all needed instructions to fill a new remote `array` with the default values),
3. And then it sets the actual values in the `array`.

While this works, instead of that, we could create a new helper to fill the type based on the context:
1. Local context -- just set all the values in the local `tuple`,
2. Remote context -- create a new remote `array` and `Append` all the local elements to it.

At least that way, the bytecode generated out of it would be smaller thanks to not emitting instructions for both `Append` and `Set` for every constructor argument.

It diverges from our "usual" way of doing it, but maybe that's the way. Thoughts?

Reply:
I'd support this change. I think erroring on the side of reducing remote instructions (as long as it doesn't make the code unreadable or anything) is fine by me.

Reply:
Hmm... I did that in a few different ways (including the one that is out now), but all of them seem somehow... ugly to me.

Thoughts? :)

Reply:
I don't know, seems fine to me. My main suggestions would be to add explanatory comments to each of the new methods and to maybe rename `ToRemoteArray` on the delegator to `ItemsToRemoteArray`. I feel like it would naturally look different, since there isn't explicit remote operation support for efficient creation of a remote tuple. But also it doesn't save _that_ many instructions (assuming we're not making gigantic tuples), so I'm also fine with the other way if you think this makes the code harder to understand.

Reply:
Hmm... The name change is fair. Some comments around methods are definitely needed, so it's a good suggestion. I also think the "appending" methods should probably be made part of the `impl` namespace instead.

As for not saving instructions, that is true. Which is why I am OK with thinking we can roll back to the previous version. :)

What stops me from doing that is more about the fact that:
1. The code is already there,
2. It is more efficient,
3. Tests cover it.

I just want to convince myself that we are not introducing a "bad" pattern. :)

Reply:
Yeah, I think it's cool :). UiaTuple is a special case anyway kinda since it isn't a wrapper for an equivalent remote type like other types are, it's just a convenience helper for strongly-typed variant arrays basically. So I'm definitely cool with the change you made.

Reply:
Yes. If it was not about non-empty default values, we would be good. As in, this can be implemented for `UiaArray` in a much simpler way just because the default array state is an empty array. Not really the case here as the number of elements for a tuple is defined by the number of field-values (and therefore the only time tuples are "empty" is when they have no fields/values - and that is not a useful scenario).

Either way, I think we are good here. Thanks for the public brainstorming. :)

---
In reply to: [505104210](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r505104210) [](ancestors = 505104210)